### PR TITLE
[WIP] Refactoring

### PR DESCRIPTION
**Pull request by florianv**
TODO:

- [x] finish to implement historical rates where possible
- [x] improve CentralBankOfRepublicTurkeyProvider
- [x] see about renaming classes / method with better names
- [ ] review caching using exchange query signatures
- [x] review exceptions

### CODE REVIEW

**Comment by Nyholm** [comment]
Could we also consider #57 and #25?

**Comment by florianv** [comment]
@Nyholm I merged / answered on them.

About this PR it's more an experiment to try to add support to historical rates, etc.

**Comment by Nyholm** [comment]
I like the concept of a query. It proves for better extensibility.

File: src/HistoricalExchangeQuery.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+     *
+     * @param CurrencyPairInterface $currencyPair
+     * @param \DateTime             $date
+     * @param array                 $options
+     */
+    public function __construct(CurrencyPairInterface $currencyPair, \DateTime $date, array $options = [])
+    {
+        parent::__construct($currencyPair, $options);
+
+        $this->date = $date;
```
Comment (Nyholm) [inline src/HistoricalExchangeQuery.php]: You should clone the date here or use immutable date

File: src/HistoricalExchangeQuery.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    private $date;
+
+    /**
+     * Creates a new query.
+     *
+     * @param CurrencyPairInterface $currencyPair
+     * @param \DateTime             $date
+     * @param array                 $options
+     */
+    public function __construct(CurrencyPairInterface $currencyPair, \DateTime $date, array $options = [])
```
Comment (Nyholm) [inline src/HistoricalExchangeQuery.php]: Use DateTimeInterface

File: src/HistoricalExchangeQueryInterface.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Contract for historical exchange queries.
+ *
+ * @author Florian Voutzinos <florian@voutzinos.com>
+ */
+interface HistoricalExchangeQueryInterface extends ExchangeQueryInterface
+{
+    /**
+     * Gets the date.
+     *
+     * @return \DateTime
```
Comment (Nyholm) [inline src/HistoricalExchangeQueryInterface.php]: \DateTimeInterface

File: src/Provider/AbstractProvider.php [hunk 4]
```diff
@@ -37,8 +37,9 @@
     /**
      * @param HttpClient|null     $httpClient
      * @param RequestFactory|null $requestFactory
+     * @param array               $options
      */
-    public function __construct(HttpClient $httpClient = null, RequestFactory $requestFactory = null)
+    public function __construct(HttpClient $httpClient = null, RequestFactory $requestFactory = null, array $options = [])
```
Comment (Nyholm) [inline src/Provider/AbstractProvider.php]: Is this WIP? The $options are never used

Comment (florianv) [inline src/Provider/AbstractProvider.php]: Yes

Comment (Nyholm) [inline src/Provider/AbstractProvider.php]: Let me know when the PR is ready for a final review. Im happy to help

File: src/HistoricalExchangeQuery.php (hunk 2 above)
Comment (florianv) [inline src/HistoricalExchangeQuery.php]: Agreed

File: src/Provider/AbstractProvider.php (hunk 4 above)
Comment (florianv) [inline src/Provider/AbstractProvider.php]: Thanks :) Sure, I will let you know.

I'm trying to implement historical rates and the missing features while making the architecture more robust and easier to extend for 3.x versions.

I plan to release a 3.0 beta for a week or so to make sure we have all needed features and it's well tested before releasing.

**Comment by florianv** [comment]
I close it in favor of https://github.com/florianv/swap/pull/66

**Pull request closed without merging**